src/tclint/symbol_table.py

```diff
@@ -0,0 +1,77 @@
+import logging
+from collections import defaultdict
+from typing import List, DefaultDict
+
+from tclint.syntax_tree import BareWord, Visitor, Command, Node
+
+
+class SymbolTable:
+    """Holds a symbol table (links symbols to nodes)."""
+
+    def __init__(self):
+        self.proc_def: DefaultDict[str, list[Node]] = defaultdict(list)
+
+    def add_proc_definition(self, command: Command) -> None:
+        """Add definition of procedure"""
+        # command holds the "proc" keyword, so the proc name is 1st argument
+        proc_name_node = command.args[0]
+        proc_name = proc_name_node.contents
+        assert proc_name is not None
+        logging.debug(
+            f"Definition of proc '{proc_name}' at {proc_name_node._pos_str()}"
+        )
+        self.proc_def[proc_name].append(proc_name_node)
+
+    def lookup_proc_definitions(self, symbol_text: str) -> List[Node]:
+        """Lookup definitions of the procedure pointed at by node"""
+        if symbol_text is None or symbol_text not in self.proc_def:
+            return []
+        return self.proc_def[symbol_text]
+
+
+class SymbolTableBuilder(Visitor):
+    """Builds a symbol table."""
+
+    def __init__(self):
+        self.table = SymbolTable()
+
+    def build(self, tree) -> SymbolTable:
+        """Run the builder visitor through the syntax tree, building a table."""
+        tree.accept(self, recurse=True)
+        return self.table
+
+    def visit_command(self, command: Command) -> None:
+        if command.routine.contents == "proc":
+            self.table.add_proc_definition(command)
+
+
+def get_symbol_text(node: Node) -> str | None:
+    """Get symbol text to look for from the node under the cursor"""
+    # TODO: where should this code live? it's not really part of a symbol table, maybe it belongs closer to the LSP stuff
+    # "proc foo" with cursor over "proc"
```

Comment:
I think I have another plan that will simplify this (and eliminate the need for the Node API prev/next tweak).

I started thinking about this in the last PR, but I'd like to address this by adding a node type called `CommandRef`. This would wrap the first argument to a `Command` node (rather than a `BareWord`), and then your logic knows any instance of a `CommandRef` is a reference to a `proc` without having to check its predecessors.

I think encoding this information in the syntax tree might be convenient for other things in the future. 

If you're interested in trying this syntax tree addition in a follow-up PR let me know, otherwise I'm happy to make some time to do it myself.

Reply:
Agree that it's good to make the syntax tree smarter. If you like to do it pls go ahead.

Will this new node include both `proc` keyword and function name? (i.e. `pos` and `end_pos` will span the area of both words).

Reply:
> Will this new node include both proc keyword and function name? (i.e. pos and end_pos will span the area of both words).

I wasn't planning on it, looking at the Python LSP for example it doesn't seem like the function keyword actually triggers a go-to def or reference for the associated function. Would you expect different behavior?

<img width="333" height="83" alt="Screenshot 2025-08-04 at 8 51 51 PM" src="https://github.com/user-attachments/assets/0937e11d-80fe-4b17-822f-90f9e360a641" />